common-pg/src/main/scala/com/socrata/pg/analyzer2/rollup/RollupExact.scala

```diff
@@ -0,0 +1,814 @@
+          log.debug("Bailing because of DISTINCT ON mismatch")
+          return None
+        }
+
+        // ORDER BY needs to be isomorphic because DISTINCT ON picks
+        // the first row
+        if(!whereishIsIsomorphic || !orderByIsIsomorphic) {
+          log.debug(s"Bailing because DISTINCT ON but $whereName or ORDER BY are not isomorphic")
+          return None
+        }
+      case _ =>
+        log.debug("Bailing because of a distinctiveness mismatch")
+        return None
+      }
+    val newDistinctiveness = rewriteDistinctiveness(sDistinctiveness, rewriteInTerms).getOrElse {
+      log.debug("Bailing because failed to rewrite distinctiveness")
+      return None
+    }
+
+    val newSelectList: OrderedMap[AutoColumnLabel, NamedExpr] =
+      OrderedMap() ++ sSelectList.iterator.mapFallibly(rewriteInTerms.rewriteSelectListEntry(_)).getOrElse {
+        log.debug("Bailing because failed to rewrite the select list")
+        return None
+      }
+
+    val (newLimit, newOffset) =
+      combineLimitOffset(sLimit, sOffset, cLimit, cOffset, orderByIsIsomorphic).getOrElse {
+        return None
+      }
+
+    Some(
+      Select(
+        distinctiveness = newDistinctiveness,
+        selectList = newSelectList,
+        from = rewriteInTerms.newFrom,
+        where = newWhere,
+        groupBy = Nil,
+        having = None,
+        orderBy = newOrderBy,
+        limit = newLimit,
+        offset = newOffset,
+        search = None,
+        hint = sHint
+      )
+    )
+  }
+
+  private def sameSelectList(s: OrderedMap[AutoColumnLabel, NamedExpr], c: OrderedMap[AutoColumnLabel, NamedExpr], isoState: IsoState): Boolean = {
+    // we don't care about order or duplicates, but we do care that
+    // all exprs in s are isomorphic to exprs in c, and vice-versa.
+    val sExprs = s.values.map(_.expr).toVector
+    val cExprs = c.values.map(_.expr).toVector
+
+    sameExprsIgnoringOrder(sExprs, cExprs, isoState)
+  }
+
+  private def sameExprsIgnoringOrder(s: Seq[Expr], c: Seq[Expr], isoState: IsoState): Boolean = {
+    def isSubset(as: Seq[Expr], bs: Seq[Expr], isoState: IsoState): Boolean = {
+      as.forall { a =>
+        bs.exists { b => a.isIsomorphic(b, isoState) }
+      }
+    }
+
+    isSubset(s, c, isoState) && isSubset(c, s, isoState.reverse)
+  }
+
+  private def rewriteDistinctiveness(distinct: Distinctiveness, rewriteInTerms: RewriteInTerms, needsMerge: Boolean = false): Option[Distinctiveness] = {
+    distinct match {
+      case Distinctiveness.Indistinct() | Distinctiveness.FullyDistinct() =>
+        Some(distinct)
+      case Distinctiveness.On(exprs) =>
+        exprs.mapFallibly(rewriteInTerms.rewrite(_, needsMerge)).map(Distinctiveness.On(_))
+    }
+  }
+
+  private def rewriteAggregatedOnUnaggregated(select: Select, candidate: Select, rewriteInTerms: RewriteInTerms): Option[Select] = {
```

Comment:
Hrm. Maybe
```suggestion
  private def rewriteAggregatedToUnaggregated(select: Select, candidate: Select, rewriteInTerms: RewriteInTerms): Option[Select] = {
```
? Not sure if there's a specific meaning to `on` I'm unaware of.

Reply:
I won't change either; the precedence is different.  We're rewriting an aggregated query on an unaggregated rollup here.  In the other case, the "one to one" means "a row of rollup maps to a row of query" (i.e., there is no further grouping happening)